R/parse-body.R

```diff
@@ -191,12 +191,22 @@ parser_multi <- function() {
     boundary <- stri_match_first_regex(content_type, "boundary=([^; ]{2,})", case_insensitive = TRUE)[,2]
     toparse <- parse_multipart(value, boundary)
     # content-type detection
-    lapply(toparse, function(x) {
+    args <- lapply(toparse, function(x) {
       if (!is.null(x$filename)) {
         x$content_type <- getContentType(tools::file_ext(x$filename))
       }
       parseRaw(x)
     })
+    # combine together args that share the same name
+    keys <- unique(names(args))
+    for (k in keys) {
+      m <- which(names(args)==k)
+      if (isTRUE(length(m) > 1)) {
+        args[[m[1]]] <- unname(args[m])
+        args[m[-1]] <- NULL
+      }
+    }
```

Comment:
Can this use the query string "same name" combiner?  Let's make that function a helper method that can be used here.

Reply:
Not in this form it is at the moment. It will have to be reworked. Also I,m worried about large copies of object in memory.

Reply:
> Not in the form it is at the moment. It will have to be reworked. Also I,m worried about large copies of object in memory.

At least in r >= 3.5 (I believe), this should not be a big concern. 

See examples of `pryr::object_size()`

```r
pryr::object_size(ggplot2::diamonds)
#> 3.46 MB
pryr::object_size(list(ggplot2::diamonds, ggplot2::diamonds, ggplot2::diamonds))
#> 3.46 MB

```


> I think it would have to be two different code path as unlisting a list of raw bytes just gives you a large raw vector.


Fair. I wouldn't mind adding a `call_unlist = TRUE` parameter to the helper method.  

The part I'm concerned about is how keys are removed from the args and how they are collected.  The query string method has been tested and I'd like to leverage it if we can.

--------------

Something like...

```r
combine_keys <- function(obj, call_unlist = TRUE) {
  keys <- names(obj)
  vals <- unname(obj)
  unique_keys <- unique(keys)

  if (length(unique_keys) == length(keys)) {
    return(obj)
  }


  cleanup_item <- 
    if (isTRUE(call_unlist)) {
      function(x) { 
        unname(unlist(x))
      }
    } else {
      unname
    }

  vals <-
    if (length(unique_keys) > 150) {
      lapply(split(vals, keys), function(items) {
        cleanup_item(items)
      })
    } else {
      # n < 150
      lapply(unique_keys, function(key) {
        cleanup_item(vals[keys == key])
      })
    }
  names(vals) <- unique_keys
  vals
}
```

```r
test_val <- replicate(5, replicate(10, charToRaw(sample(letters, 1))), simplify = FALSE) %>% setNames(c("a", "b", "a", "b", "a")) %>% print()
#> $a
#>  [1] 6f 66 6b 6d 73 79 63 67 66 6f
#> 
#> $b
#>  [1] 6a 62 63 6d 62 76 71 78 64 78
#> 
#> $a
#>  [1] 61 62 65 70 6b 67 7a 62 63 73
#> 
#> $b
#>  [1] 62 74 74 6b 64 6a 68 6b 75 74
#> 
#> $a
#>  [1] 64 67 6a 6d 75 72 76 77 6a 67
#> 
#> 
combine_keys(test_val, TRUE)
#> $a
#>  [1] 6f 66 6b 6d 73 79 63 67 66 6f 61 62 65 70 6b 67 7a 62 63 73 64 67 6a 6d 75 72 76 77 6a 67
#> 
#> $b
#>  [1] 6a 62 63 6d 62 76 71 78 64 78 62 74 74 6b 64 6a 68 6b 75 74
#> 
#> 
combine_keys(test_val, FALSE)
#> $a
#> $a[[1]]
#>  [1] 6f 66 6b 6d 73 79 63 67 66 6f
#> 
#> $a[[2]]
#>  [1] 61 62 65 70 6b 67 7a 62 63 73
#> 
#> $a[[3]]
#>  [1] 64 67 6a 6d 75 72 76 77 6a 67
#> 
#> 
#> $b
#> $b[[1]]
#>  [1] 6a 62 63 6d 62 76 71 78 64 78
#> 
#> $b[[2]]
#>  [1] 62 74 74 6b 64 6a 68 6b 75 74
```

Reply:
Yeah I don't mind doing that somewhere else if it handle memory, I'll fiddle with it. and append this one.

Reply:
What do you think about two separate code path? keeping the lapply transform the args structure by listing every element. Even those that are not duplicate when a duplicate is present in the list.

```r
sample1 <- list(a = women, b = 1:4, c = mtcars)
sample2 <- list(a = women, b = 1:4, a = mtcars, c=mtcars)
expect_equal(combine_keys(sample1, FALSE), sample1)
# This one will fail
expect_equal(combine_keys(sample2, FALSE), list(a = list(women, mtcars), b = 1:4, c=mtcars))

# It would return
# list(a = list(women, mtcars), b = list(1:4), c=list(mtcars))
```

Reply:
This would only apply for `parser_multi`?  I think having each key point to a list of items is a great idea.

Makes everything consistent to parse. Good catch!

Yes. Let's makes it return like:
> ```
> # list(a = list(women, mtcars), b = list(1:4), c=list(mtcars))
> ```

Reply:
I meant adding the edge case to the combine function.
You can check the PR files now.

Returning always a list for each key would make behavior inconsistent.

```R
#* @param dt:df
function(dt = mtcars) {
   class(dt)
   # data.frame
}

#* @param dt:df
#* @param files:[file]
function(dt = mtcars, files) {
   class(dt)
   # list
   class(dt[[1]])
   # data.frame
   class(files)
   # list
}
```

Reply:
> Makes everything consistent to parse. Good catch!
You have a point there, if you send in only one file to an endpoint that can accept multiple files, you would not receive a list of file. Back to drawing board.